### PR TITLE
Fix unsigned integer bitnot

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**
fixes https://github.com/lcompilers/lpython/issues/2054

### CODE REVIEW

**Comment by Shaikh-Ubaid** [comment]
In the commit [lpython.py: Use numpy functions](https://github.com/lcompilers/lpython/pull/2061/commits/465bb814cd4ad76b5c24ad1829800ce43825a018), I replaced custom functions with `numpy` functions. This way we need not implement every operator for custom Unsigned Integer class. A drawback here seems to be that everytime `lpython` is imported, it loads `numpy` (and thus could affect performance of `CPython`).

Please share if it is fine.

**Comment by Shaikh-Ubaid** [comment]
Ready.

**Comment by certik** [comment]
I would only depend on CPython for now by default. Of course if you use NumPy then you need NumPy, but I would not require it if you don't use NumPy.

**Comment by certik** [comment]
I think this looks great. Do you mind splitting the assert improvement into a separate PR? It will be much easier to review.

**Comment by Shaikh-Ubaid** [comment]
Ready.

File: integration_tests/bindpy_01.py [hunk 1]
```diff
@@ -1,11 +1,11 @@
 from lpython import i32, i64, u32, u64, f32, f64, pythoncall
 
 @pythoncall(module = "bindpy_01_module")
-def add_ints(a: i32, b: i64, c: u32, d: u64) -> i64:
+def add_ints(a: u32, b: u32, c: u32, d: u32) -> u64:
     pass
 
 @pythoncall(module = "bindpy_01_module")
-def multiply_ints(a: i32, b: i64, c: u32, d: u64) -> i64:
+def multiply_ints(a: u32, b: u32, c: u32, d: u32) -> u64:
```
Comment (certik) [inline integration_tests/bindpy_01.py:8]: Are the signed integers tested somewhere? It seems we are now only testing unsigned integers, but we need to test both signed and unsigned.

**Comment by certik** [comment]
I think this PR is great now, except that it seems the signed integer tests were removed --- if you can test both signed and unsigned, then we can merge it.

File: integration_tests/bindpy_01.py [hunk 2]
```diff
@@ -1,11 +1,19 @@
 from lpython import i32, i64, u32, u64, f32, f64, pythoncall
 
 @pythoncall(module = "bindpy_01_module")
-def add_ints(a: i32, b: i64, c: u32, d: u64) -> i64:
+def add_ints(a: i32, b: i32, c: i32, d: i32) -> i64:
```
Comment (certik) [inline integration_tests/bindpy_01.py:4]: Is i64 not tested now?

File: integration_tests/bindpy_01.py [hunk 3]
```diff
@@ -1,11 +1,19 @@
...
     pass
 
 @pythoncall(module = "bindpy_01_module")
-def multiply_ints(a: i32, b: i64, c: u32, d: u64) -> i64:
+def multiply_ints(a: i32, b: i32, c: i32, d: i32) -> i64:
+    pass
+
+@pythoncall(module = "bindpy_01_module")
+def add_unsigned_ints(a: u32, b: u32, c: u32, d: u32) -> u64:
```
Comment (certik) [inline integration_tests/bindpy_01.py:12]: Let's test u64 too.

**Comment by certik** [comment]
I would say in this PR make sure all previous tests are left, and just add some unsigned tests.

Then in the next PR, let's test rigorously i8, i16, i32, i64 and u8, u16, u32, u64 to make sure they all work.

**Comment by Shaikh-Ubaid** [comment]
>Is i64 not tested now?
>Let's test u64 too.
>I would say in this PR make sure all previous tests are left, and just add some unsigned tests.

Yes, `i64` is not tested as a direct parameter. It might be tested as an array of `i64` (in some other test). 

These tests were added to test initial support of `cpython` interop. I would not make the test big. I would add it as a separate-test / new-test-file (if needed). I think simple and smaller tests are much simpler to debug and figure the failure. Testing all the types into one file could lead to tricky debugging, I think.

These types will be tested thoroughly as we add more support for CPython-Interop (For example the pass at https://github.com/lcompilers/lpython/issues/2050). When I added this test initially, I willingly kept the test case slightly less robust (and even slightly odd, for example multiplying/adding variables of different types together) so that we can have some freedom (or thinking/discussion opportunities) in the design/structure of CPython Interop. 

We are needing to update this test `bindpy_01.py` since the UnsignedInteger class defined in `lpython.py` (for `CPython`) supports operations only its own instances. This was needed because after any operation we need the unsigned integer value to be in its valid range (as per its kind). 

Currently, the updated test tests integers (only `i32`) and unsigned integers (only `u32`) which seems good. We can add more tests in a separate file, I think.

Please share what we should do.

**Review by certik — APPROVED** [review]

**Comment by certik** [comment]
I think this is fine. We should now make sure all combinations are tested.

**Comment by certik** [comment]
We made a mistake here. The UnsignedIntegerBitNot wraps around, and thus does NOT correspond do the CPython's `~` operator, which does not wrap around. A solution is here:

https://github.com/lcompilers/lpython/issues/2173

We need to create a new intrinsic function `bitnot(u)` that maps to `UnsignedIntegerBitNot`, and then we must forbid `~u` applied to unsigned integers. Then we remove the UnsignedInteger class in emulation mode. That brings us back to be compatible with CPython.

**Pull request merged**